snowfall/models/tdnnf.py

```diff
@@ -0,0 +1,399 @@
+        super().__init__()
+        self.affine = nn.Linear(in_features=small_dim, out_features=big_dim)
+        self.batchnorm1 = nn.BatchNorm1d(num_features=big_dim, affine=False)
+        self.linear = OrthonormalLinear(dim=big_dim,
+                                        bottleneck_dim=small_dim,
+                                        kernel_size=1)
+        self.batchnorm2 = nn.BatchNorm1d(num_features=small_dim, affine=False)
+
+    def forward(self, x):
+        # x is [N, C, T]
+        x = x.permute(0, 2, 1)
+
+        # at this point, x is [N, T, C]
+
+        x = self.affine(x)
+        x = F.relu(x)
+
+        # at this point, x is [N, T, C]
+
+        x = x.permute(0, 2, 1)
+
+        # at this point, x is [N, C, T]
+
+        x = self.batchnorm1(x)
+
+        x = self.linear(x)
+
+        x = self.batchnorm2(x)
+
+        return x
+
+
+class TDNN(nn.Module):
+    '''
+    This class implements the following topology in kaldi:
+      relu-batchnorm-dropout-layer name=tdnn1 dropout-per-dim-continuous=true dim=1024
+    '''
+
+    def __init__(self, input_dim, hidden_dim):
+        super().__init__()
+        # affine conv1d requires [N, C, T]
+        self.affine = nn.Conv1d(in_channels=input_dim,
+                                out_channels=hidden_dim,
+                                kernel_size=1)
+
+        # tdnn1_batchnorm requires [N, C, T]
+        self.batchnorm = nn.BatchNorm1d(num_features=hidden_dim,
+                                        affine=False)
+
+        self.dropout = SharedDimScaleDropout(dim=2)
+
+    def forward(self, x, dropout=0.):
+        # input x is of shape: [batch_size, feat_dim, seq_len] = [N, C, T]
+        x = self.affine(x)
+        x = F.relu(x)
+        x = self.batchnorm(x)
+        x = self.dropout(x, alpha=dropout)
+        # return shape is [N, C, T]
+        return x
+
+
+class FactorizedTDNN(nn.Module):
+    '''
+    This class implements the following topology in kaldi:
+      tdnnf-layer name=tdnnf2 $tdnnf_opts dim=1024 bottleneck-dim=128 time-stride=1
+    References:
+        - http://danielpovey.com/files/2018_interspeech_tdnnf.pdf
+        - ConstrainOrthonormalInternal() from
+          https://github.com/kaldi-asr/kaldi/blob/master/src/nnet3/nnet-utils.cc#L982
+    '''
+
+    def __init__(self,
+                 dim,
+                 bottleneck_dim,
+                 kernel_size,
+                 subsampling_factor,
+                 bypass_scale=0.66,
+                 cnn_padding=1):
+        super().__init__()
+
+        assert abs(bypass_scale) <= 1
+
+        self.bypass_scale = bypass_scale
+
+        self.s = subsampling_factor
+
+        # linear requires [N, C, T]
+        self.linear = OrthonormalLinear(dim=dim,
+                                        bottleneck_dim=bottleneck_dim,
+                                        kernel_size=kernel_size)
+
+        # affine requires [N, C, T]
+        # WARNING(fangjun): we do not use nn.Linear here
```

Comment:
mm.  this is kind of inefficient.  It's 3x more efficient, for later layers, if instead of making use of `stride`, we just subsample by 3x after the last layer that has a stride of 1 in the Kaldi config.

Kaldi implicitly subsamples.

Reply:
.. we can merge anyway though, if it might take some time to address that.

Reply:
> It's 3x more efficient, for later layers, if instead of making use of stride, we just subsample by 3x after the last layer that has a stride of 1 in the Kaldi config.

Currently, it uses stride==3 for the last layer that has a stride of 1 in kaldi config.
Does it achieve the same goal?

Reply:
oh sorry, no, that's right, it's dilation that we shoudn't be using, not stride.